rust/arrow/src/array/array.rs

```diff
@@ -976,6 +1006,53 @@ impl BinaryArray {
         }
     }
 
+    /// Returns the offset for the element at index `i`.
+    ///
+    /// Note this doesn't do any bound checking, for performance reason.
+    #[inline]
+    pub fn value_offset(&self, i: usize) -> i32 {
+        self.value_offset_at(self.data.offset() + i)
+    }
+
+    /// Returns the length for the element at index `i`.
+    ///
+    /// Note this doesn't do any bound checking, for performance reason.
+    #[inline]
+    pub fn value_length(&self, mut i: usize) -> i32 {
+        i += self.data.offset();
+        self.value_offset_at(i + 1) - self.value_offset_at(i)
+    }
+
+    /// Returns a clone of the value offset buffer
+    pub fn value_offsets(&self) -> Buffer {
+        self.data.buffers()[0].clone()
+    }
+
+    /// Returns a clone of the value data buffer
+    pub fn value_data(&self) -> Buffer {
+        self.data.buffers()[1].clone()
+    }
+
+    #[inline]
+    fn value_offset_at(&self, i: usize) -> i32 {
+        unsafe { *self.value_offsets.get().offset(i as isize) }
+    }
+}
+
+impl StringArray {
+    /// Returns the element at index `i` as a byte slice.
+    pub fn value(&self, i: usize) -> &[u8] {
+        assert!(i < self.data.len(), "StringArray out of bounds access");
+        let offset = i.checked_add(self.data.offset()).unwrap();
```

Comment:
Shouldn't we be checking that `i + offset < len` instead of just `i`?

Also, returning a `Result` is probably unnecessary but maybe use `expect` instead of `unwrap`.